logger/storage/file/adapter.go

```diff
@@ -0,0 +1,130 @@
+package file
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path"
+	"strconv"
+	"strings"
+	"sync"
+)
+
+type adapter struct {
+	logRoot string
+	files   map[string]*os.File
+	mutex   sync.Mutex
+}
+
+// NewStorageAdapter returns a pointer to a new instance of a file-based storage.Adapter.
+func NewStorageAdapter(logRoot string) (*adapter, error) {
+	src, err := os.Stat(logRoot)
+	if err != nil {
+		return nil, fmt.Errorf("Directory %s does not exist", logRoot)
+	}
+	if !src.IsDir() {
+		return nil, fmt.Errorf("%s is not a directory", logRoot)
+	}
+	return &adapter{logRoot: logRoot, files: make(map[string]*os.File)}, nil
+}
+
+// Write adds a log message to to an app-specific log file
+func (a *adapter) Write(app string, message string) error {
+	// Check first if we might actually have to add to the map of file pointers so we can avoid
+	// waiting for / obtaining a lock unnecessarily
+	f, ok := a.files[app]
+	if !ok {
+		// Ensure only one goroutine at a time can be adding a file pointer to the map of file
+		// pointers
+		a.mutex.Lock()
+		defer a.mutex.Unlock()
+		f, ok = a.files[app]
+		if !ok {
+			var err error
+			f, err = a.getFile(app)
+			if err != nil {
+				return err
+			}
+			a.files[app] = f
+		}
+	}
+	if _, err := f.WriteString(message + "\n"); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Read retrieves a specified number of log lines from an app-specific log file
+func (a *adapter) Read(app string, lines int) ([]string, error) {
+	if lines <= 0 {
+		return []string{}, nil
+	}
+	filePath := a.getFilePath(app)
+	exists, err := fileExists(filePath)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, fmt.Errorf("Could not find logs for '%s'", app)
+	}
+	logBytes, err := exec.Command("tail", "-n", strconv.Itoa(lines), filePath).Output()
+	if err != nil {
+		return nil, err
+	}
+	logStrs := strings.Split(string(logBytes), "\n")
+	return logStrs[:len(logStrs)-1], nil
+}
+
+func (a *adapter) Destroy(app string) error {
+	filePath := a.getFilePath(app)
```

Comment:
I think there are race conditions related to the file. It seems that there are a number of places where between fileExists and another operation, the file can be created or destroyed. Calls to Write and Destroy at around the same time could cause unpredictable states, since Destroy has no mutex hold.

And I might be misreading, but it looks like a file can be destroyed, but still be in `a.files`.